### PR TITLE
Add method to ED1d for calculating membrane resistance as a function of ion concentration

### PR DESCRIPTION
**Pull request by lbibl**
## Fixes/Resolves:

(replace this with the issue # fixed or resolved, if no issue exists then a brief statement of what this PR does)

## Summary/Motivation:

- Add method to ED1d for calculating membrane resistance as a function of ion concentration
- Adapt all tests and relevant flowsheet to this change. 
- Minor improvement on the single-pass Ed flowsheet for stability. 

## Changes proposed in this PR:
- As above.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/flowsheets/electrodialysis/electrodialysis_1stack_conc_recirc_ui.py
Comment (johnson12742) [inline watertap/flowsheets/electrodialysis/electrodialysis_1stack_conc_recirc_ui.py]: Typo in resistance

File: watertap/unit_models/electrodialysis_1D.py
Comment (johnson12742) [inline watertap/unit_models/electrodialysis_1D.py]: 'High electrolyte' confused me. I assumed it meant in Concentrate channel. This really means resistance at infinite concentration, right?

File: watertap/unit_models/electrodialysis_1D.py
Comment (johnson12742) [inline watertap/unit_models/electrodialysis_1D.py:812]: The whole system is electrically neutral but does the diluate stay electrically neutral at every 'x'?

Comment (lbibl) [reply watertap/unit_models/electrodialysis_1D.py:812]: yes

File: watertap/unit_models/electrodialysis_1D.py
Comment (lbibl) [reply watertap/unit_models/electrodialysis_1D.py]: yes. I hesitated about using the infinite word as it's not practically meaningful.  but I'll make an update.

Comment (lbibl) [reply watertap/unit_models/electrodialysis_1D.py]: this is updated to "infinity-approximated electrolyte concentration"

**Review by johnson12742 — COMMENTED** [review]
Which references talks about the membrane resistance as a function of ion concentration? I quickly glanced through a few and didn't find one that talks about this function.

I was recommended the Galama paper for it (https://doi.org/10.1016/j.memsci.2014.05.046)
Also, this is an empirical relationship (is that correct?). Maybe make a note of that. 

Everything else looks good.

**Comment by hunterbarber** [comment]
@lbibl Can you clarify on the meaning of the `spacer_conductivity_coefficient`? I'm a little confused while going through the review.

**Comment by lbibl** [comment]
> @lbibl Can you clarify on the meaning of the `spacer_conductivity_coefficient`? I'm a little confused while going through the review.

Added notes in doc: "We used a coefficient multiplied by the solution conductance, denoted by :math:\sigma, to account for the spacer's conductance shadowing effect."   There are cases where the shadowing effect is treated as a coefficient to resistance rather than conductance.  This would be a clearer naming with the added notation.

**Comment by lbibl** [comment]
> Which references talks about the membrane resistance as a function of ion concentration? I quickly glanced through a few and didn't find one that talks about this function.
> 
> I was recommended the Galama paper for it (https://doi.org/10.1016/j.memsci.2014.05.046) Also, this is an empirical relationship (is that correct?). Maybe make a note of that.
> 
> Everything else looks good.

Updates on references underway.

**Review by johnson12742 — APPROVED** [review]

File: watertap/unit_models/electrodialysis_1D.py
Comment (hunterbarber) [inline watertap/unit_models/electrodialysis_1D.py:543]:
```suggestion
            doc='The porosity of spacer in the ED channels. This is also referred to elsewhere as "void fraction"',
```

File: watertap/unit_models/electrodialysis_1D.py
Comment (hunterbarber) [inline watertap/unit_models/electrodialysis_1D.py:610]:
```suggestion
            doc="Coefficient of membrane areal resistance to 1/c, where c is the diluate electrolyte concentration",
```

File: watertap/unit_models/electrodialysis_1D.py
Comment (hunterbarber) [inline watertap/unit_models/electrodialysis_1D.py]:
```suggestion
                "Total electrical power consumption (W)": self.diluate.power_electrical_x[
```

File: watertap/unit_models/tests/test_electrodialysis_1D.py
Comment (hunterbarber) [inline watertap/unit_models/tests/test_electrodialysis_1D.py]:
```suggestion
            perform_dict["vars"]["Total electrical power consumption (W)"]
```

File: watertap/unit_models/tests/test_electrodialysis_1D.py
Comment (hunterbarber) [inline watertap/unit_models/tests/test_electrodialysis_1D.py]:
```suggestion
            perform_dict["vars"]["Total electrical power consumption (W)"]
```

File: watertap/unit_models/tests/test_electrodialysis_1D.py
Comment (hunterbarber) [inline watertap/unit_models/tests/test_electrodialysis_1D.py]:
```suggestion
            perform_dict["vars"]["Total electrical power consumption (W)"]
```

File: watertap/unit_models/tests/test_electrodialysis_1D.py
Comment (hunterbarber) [inline watertap/unit_models/tests/test_electrodialysis_1D.py]:
```suggestion
            perform_dict["vars"]["Total electrical power consumption (W)"]
```

File: watertap/unit_models/tests/test_electrodialysis_1D.py
Comment (hunterbarber) [inline watertap/unit_models/tests/test_electrodialysis_1D.py]:
```suggestion
            perform_dict["vars"]["Total electrical power consumption (W)"]
```

File: watertap/unit_models/tests/test_electrodialysis_1D.py
Comment (hunterbarber) [inline watertap/unit_models/tests/test_electrodialysis_1D.py]:
```suggestion
            perform_dict["vars"]["Total electrical power consumption (W)"]
```

File: watertap/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (hunterbarber) [inline watertap/flowsheets/electrodialysis/electrodialysis_1stack.py:272]: I would summarize the changes made in the PR in the "Changes proposed in this PR:" section for tractability, and to clarify why this is changing

File: watertap/flowsheets/electrodialysis/electrodialysis_1stack_conc_recirc_ui.py
Comment (hunterbarber) [inline watertap/flowsheets/electrodialysis/electrodialysis_1stack_conc_recirc_ui.py]: Do you need the add the `spacer_conductivity_coefficient` as a UI variable?

File: docs/technical_reference/unit_models/electrodialysis_1D.rst
Comment (hunterbarber) [inline docs/technical_reference/unit_models/electrodialysis_1D.rst]:
```suggestion
   "Membrane areal resistance ion concentration-independent constant", ":math:`r_{const}`", "membrane_areal_resistance_const", "['cem', 'aem']", ":math:`\Omega m^2`", 2
```

File: docs/technical_reference/unit_models/electrodialysis_1D.rst
Comment (hunterbarber) [inline docs/technical_reference/unit_models/electrodialysis_1D.rst]:
```suggestion
   "Membrane areal resistance ion concentration-dependent coefficient", ":math:`r_{coef}`", "membrane_areal_resistance_coef", "['cem', 'aem']", ":math:`\Omega mol m^{-1}`", 2
```

**Review by hunterbarber — COMMENTED** [review]
Only minor comments that should be considered before merging.

File: watertap/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (lbibl) [reply watertap/flowsheets/electrodialysis/electrodialysis_1stack.py:272]: This change is insignificant. Both work fine.

File: watertap/flowsheets/electrodialysis/electrodialysis_1stack_conc_recirc_ui.py
Comment (lbibl) [reply watertap/flowsheets/electrodialysis/electrodialysis_1stack_conc_recirc_ui.py]: no and not all variables are subject to UI.

**Comment by lbibl** [comment]
@hunterbarber  changing "Watt" to"W" is causing several test failing because that string was used as keys in many tests consistently.  I consider this change very trivial and not worthy of the time to go through so I am dismissing it.  If you insist, please feel free to open another PR for changes.

File: docs/technical_reference/unit_models/electrodialysis_1D.rst
Comment (lbibl) [reply docs/technical_reference/unit_models/electrodialysis_1D.rst]: The original is more accurate because that is not a dimensionless constant but a physically meaningful areal resistance at infinitive electrolyte concentration.

File: docs/technical_reference/unit_models/electrodialysis_1D.rst
Comment (lbibl) [reply docs/technical_reference/unit_models/electrodialysis_1D.rst]: updated.

File: docs/technical_reference/unit_models/electrodialysis_1D.rst
Comment (hunterbarber) [reply docs/technical_reference/unit_models/electrodialysis_1D.rst]: I was originally going to just suggest "Membrane areal resistance independent _of_ ion concentration" but then was matching the variable nomenclature which uses '_const' throughout which I assumed was in reference to constant parameter (not dimensionless constant).

**Comment by hunterbarber** [comment]
> @hunterbarber changing "Watt" to"W" is causing several test failing because that string was used as keys in many tests consistently. I consider this change very trivial and not worthy of the time to go through so I am dismissing it. If you insist, please feel free to open another PR for changes.

This string has already been changed in this PR though, meaning the repetitive checks for it were already revised once. I don't think it should hold up this being merged, but if the typo and consistency change from "consumption(Watt)" to "consumption (W)" doesn't happen in this PR, I bet it won't ever get corrected.

File: watertap/flowsheets/electrodialysis/electrodialysis_1stack.py
Comment (hunterbarber) [reply watertap/flowsheets/electrodialysis/electrodialysis_1stack.py:272]: I agree this change is insignificant. It only becomes significant if a user is referencing this flowsheet for something unaware to us, and their code breaks after merging with this PR. That's why at the minimum the "Fixes", "Summary", and "Changes" sections of the PR should have information regarding the changes as a history to reference.

**Comment by lbibl** [comment]
> > @hunterbarber changing "Watt" to"W" is causing several test failing because that string was used as keys in many tests consistently. I consider this change very trivial and not worthy of the time to go through so I am dismissing it. If you insist, please feel free to open another PR for changes.
> 
> This string has already been changed in this PR though, meaning the repetitive checks for it were already revised once. I don't think it should hold up this being merged, but if the typo and consistency change from "consumption(Watt)" to "consumption (W)" doesn't happen in this PR, I bet it won't ever get corrected.

No it's not changed now.  I had to push it back.  are you referring "Watt" as a typo but it is not?

**Comment by hunterbarber** [comment]
> > > @hunterbarber changing "Watt" to"W" is causing several test failing because that string was used as keys in many tests consistently. I consider this change very trivial and not worthy of the time to go through so I am dismissing it. If you insist, please feel free to open another PR for changes.
> > 
> > 
> > This string has already been changed in this PR though, meaning the repetitive checks for it were already revised once. I don't think it should hold up this being merged, but if the typo and consistency change from "consumption(Watt)" to "consumption (W)" doesn't happen in this PR, I bet it won't ever get corrected.
> 
> No it's not changed now. I had to push it back. are you referring "Watt" as a typo but it is not?

Looking at the code difference, it looks like the addition of units into the print strings within the var dictionary was a change in this PR. The lack of a space in the spring being the typo, and using Watt here and kW, h, m, etc. elsewhere as the inconsistency.

**Review by hunterbarber — APPROVED** [review]

**Pull request merged**
